engine/consensus/approvals/chunk_collector.go

```diff
@@ -0,0 +1,72 @@
+package approvals
+
+import (
+	"sync"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// ChunkProcessingStatus is a helper structure which is used to report
+// chunk processing status to caller
+type ChunkProcessingStatus struct {
+	numberOfApprovals uint
+	approvalProcessed bool
+}
+
+// ChunkApprovalCollector implements logic for checking chunks against assignments as
+// well as accumulating signatures of already checked approvals.
+type ChunkApprovalCollector struct {
+	assignment     map[flow.Identifier]struct{} // set of verifiers that were assigned to current chunk
+	chunkApprovals *flow.SignatureCollector     // accumulator of signatures for current collector
+	lock           sync.Mutex                   // lock to protect `chunkApprovals`
+}
+
+func NewChunkApprovalCollector(assignment map[flow.Identifier]struct{}) *ChunkApprovalCollector {
+	return &ChunkApprovalCollector{
+		assignment:     assignment,
+		chunkApprovals: flow.NewSignatureCollector(),
+		lock:           sync.Mutex{},
+	}
+}
+
+// ProcessApproval performs processing and bookkeeping of single approval
+func (c *ChunkApprovalCollector) ProcessApproval(approval *flow.ResultApproval) ChunkProcessingStatus {
```

Comment:
Originally I didn't want to pass `requiredApprovalsForSealConstruction` to ChunkApprovalCollector since it was responsible only for collecting signatures, I can't remember exact reasoning, maybe it was because of testing or smth like that. If you want I can change it. As I said, originally it was meant to collect approvals for one particular chunk, without knowledge of what will happen next.

Reply:
I think the benefit of moving `requiredApprovalsForSealConstruction` into `ChunkApprovalCollector ` would be:
* Reaching the required number of approval, computing (and potentially caching the aggregated signature) would be one atomic operation. This means I am guaranteed to stop collecting approvals for a chunk as soon as I have reached the threshold. 
* Without atomicity, it is hard to argue what will precisely happen. I think the current implementation might be adding more approvals than required (under certain race conditions). 

When it comes to the protocol implementation, I tend to lean more towards implementations where thresholds are precisely defined. Though, in this special case, I don't think there is any harm in admitting a few more approvals.

Reply:
You could even take it a step further: 
```golang
func (c *ChunkApprovalCollector) ProcessApproval(approval *flow.ResultApproval) (flow.AggregatedSignature, bool)
```
thereby, you avoid another locking operation on the happy path and reduce API surface. ;-)